lib/rspec/expectations/failure_aggregator.rb

```diff
@@ -41,7 +41,11 @@ def other_errors
 
       # This method is defined to satisfy the callable interface
       # expected by `RSpec::Support.with_failure_notifier`.
-      def call(failure)
+      def call(failure, options)
+        source_id = options[:source_id]
+        return if source_id && @seen_source_ids.key?(source_id)
+
+        @seen_source_ids[source_id] = true
```

Comment:
Why not write this as a hash with an || assign, and map the failures out at the end, I think that could be neater.

```Ruby
failures[options[:source_id]] ||=
  begin
    assign_backtrace(failure) unless failure.backtrace
    failure
  end
```